syntax = "proto3";

//Clien Requests
service KeyValueStoreRPC{

  // Get Request From Client
  rpc GetRPC(StoreRequest) returns (StoreResponse);
  
  // Delete Request From Client
  rpc WriteRPC(StoreRequest) returns (StoreResponse);
}
message StoreResponse{
  bool ok=1;
  string key=2;
  string value=3;
}
message StoreRequest{
  int32 operation=1; // 1: Add | 0:drop
  string key=2;
  string value=3;
  int64 id=4;

}

//Raft Inter node RPCs
service Raft{

  //
  rpc SendLogEntry(LogRequest) returns (LogResponse);
  
  //
  rpc CommitLogEntry(CommitRequest) returns (CommitResponse);

  //
  rpc Heartbeat(HeartRequest) returns (BeatsResponse);
  
  //
  rpc VoteRPC(VoteRequest) returns(VoteResponse);
  
  //
  rpc NewLeader(LeaderChangeRequest) returns (LeaderChangeResponse);
  
  //
  rpc UpdateClusterMember(MemberRequest) returns (MemberResponse);
  
  //
  rpc ShareClusterInfo(ClusterInfo) returns (CommitResponse);
  
}

// message common_response{
//   bool success=1;
//   int32 term=2;
//   string leader_ip_port=3;

//   string value=4;
// }
message LogRequest{
  repeated StoreRequest entries=1;
  int32 term=2;
  int32 request_type=3; // 0:put, 1 delete
}
message LogResponse{
  bool success=1;
  int32 term=2;
  string leader_ip_port=3;

  string value=4;
  
}

message CommitRequest{
  int64 entry_id=1;
  bool commit=2; // 1: commit this entry, 0 discard this entry
  int32 term=3;
}
message CommitResponse{
  bool success=1 ; // 0 : discarded
}

message HeartRequest{
  int32 term=1;
}
message BeatsResponse{
  bool is_leader=1;
  int32 term=2;
  string leader_ip_port=3;
}

message VoteRequest{
  int32 term = 1;
  string ip_port = 2;
  int64 last_commit_index=3;
}
message VoteResponse{
  bool success=1;
  int32 term=2;
}

message LeaderChangeRequest{
  int32 term=1;
  string ip_port=2;
}
message LeaderChangeResponse{
  bool success=1;
  int32 term=2;
}

message MemberRequest{
  bool broadcast=1;
  string ip_port=2;
  string cluster_key=3;
}
message MemberResponse{
  bool success=1;
    int32 term=2;
  string leader_ip_port=3;

}

message ClusterInfo{
  string cluster_key=1;
  repeated string ip_port=2; 
  string leader_ip_port=3;
  int32 term=4;
}
